packages/typegpu/src/core/function/astUtils.ts

```diff
@@ -0,0 +1,28 @@
+import type { Block } from 'packages/tinyest/dist';
+import type { AnyTgpuData } from '../..';
+import type { TgslImplementation } from './fnTypes';
+
+export type Ast = {
+  argNames: string[];
+  body: Block;
+  externalNames: string[];
+};
+
+export type AstInfo = {
+  ast: Ast;
+  externals?: Record<string, unknown> | undefined;
+};
+
+export const pluginAstInfo = new WeakMap<TgslImplementation, AstInfo>();
```

Comment:
Tiny nitpick, the plugin is only responsible for inserting a call to populate this map, and the meaning is a bit lost in this context. I would make it a bit more self-documenting about its purpose, rather than its origin.

Also, narrowing the API.

```suggestion
const functionToAstMap = new WeakMap<(...args: unknown[]) => unknown, AstInfo>();

export function getPrebuiltAstFor(fn: (...args: unknown[]) => unknown): AstInfo | undefined {
  return functionToAstMap.get(fn);
}
```